<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
from config import DBNAME, USER, HOST, PASSWORD, PORT

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [58]:
# текст запроса
query_3_1 = f'''select * 
                from vacancies
'''

In [59]:
# результат запроса
df_vacancies = pd.read_sql_query(query_3_1, connection)
display(df_vacancies.shape[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\4057811422.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacancies = pd.read_sql_query(query_3_1, connection)


49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [60]:
query_3_2 = f''' select *
                 from employers
'''

In [61]:
# результат запроса
df_employers = pd.read_sql_query(query_3_2, connection)
display(df_employers.shape[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1401624884.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employers = pd.read_sql_query(query_3_2, connection)


23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [62]:
query_3_3 = f''' select *
                 from areas
'''

In [63]:
# результат запроса
df_areas = pd.read_sql_query(query_3_3, connection)
display(df_areas.shape[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1916541560.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_areas = pd.read_sql_query(query_3_3, connection)


1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [64]:
query_3_4 = f''' select *
                 from industries
'''

In [65]:
# результат запроса
df_industries = pd.read_sql_query(query_3_4, connection)
display(df_industries.shape[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\764892875.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_industries = pd.read_sql_query(query_3_4, connection)


294

***

### Выводы по предварительному анализу данных
На данном этапе предварительного аназила, мы выяснили размерность наших таблиц, а именно количество строчных данных. 

### Подведем итоги:
- Количество вакансий - 49197
- Количество работодателей - 23501
- Количество регионов - 1362
- Количество сфер деятельности в базе - 294


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [67]:
query_4_1 = f'''select 
                    a.name as area, 
                    count(*) as cnt 
                from areas as a
                join vacancies as v on a.id = v.area_id  
                group by area
                order by cnt desc
                limit 5
'''

In [68]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\234815181.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [69]:
query_4_2 = f'''select count(*)
                from vacancies 
                where (salary_from is not null) or (salary_to is not null)
'''

In [70]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
display(df.iloc[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\3341000223.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count    24073
Name: 0, dtype: int64

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [71]:
query_4_3 = f'''select 
                round(avg(salary_from)) as avg_salary_from,
                round(avg(salary_to)) as avg_salary_to
                from vacancies
'''

In [72]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\327040603.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [73]:
display(df_vacancies.columns)

Index(['id', 'name', 'key_skills', 'schedule', 'experience', 'employment',
       'salary_from', 'salary_to', 'area_id', 'employer_id'],
      dtype='object')

In [74]:
query_4_4 = ''' select 
                    schedule,
                    employment,
                    count(*) as cnt_combinations
                from vacancies
                group by 1,2 
                order by 3 desc
'''

In [75]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\3274844153.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  cnt_combinations
0        Полный день     Полная занятость             35367
1   Удаленная работа     Полная занятость              7802
2      Гибкий график     Полная занятость              1593
3   Удаленная работа  Частичная занятость              1312
4     Сменный график     Полная занятость               940
5        Полный день           Стажировка               569
6     Вахтовый метод     Полная занятость               367
7        Полный день  Частичная занятость               347
8      Гибкий график  Частичная занятость               312
9        Полный день     Проектная работа               141
10  Удаленная работа     Проектная работа               133
11     Гибкий график           Стажировка               116
12    Сменный график  Частичная занятость               101
13  Удаленная работа           Стажировка                64
14     Гибкий график     Проектная работа                18
15    Сменный график           Стажировка                12
16    Вахтовый метод     Проектная работа                 2
17    Сменный график     Проектная работа                 1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [76]:
query_4_5 = ''' select experience
                from vacancies
                group by experience  
                order by count(*)
'''

In [77]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\969833449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

### Выводы по детальному анализу вакансий

На основе данных можно сделать выводы:
- Больше всего вакансий в таких городах, как Москва, Санкт-Петербург, Минск, Новосибирск (ничего удивительного и аномального).
- Приблизительно у половины вакансий заполнено хотя бы одно из двух полей с зарплатой.
- Среднее значение для нижней и верхней границы заработной вилки составляет - 71 065 руб.  и 110 537 руб. (достаточно хорошая заработная вилка).
- Основную часть типа рабочего графика и типа трудоустройства составляют:
    - (Полный день; Полная занятость)
    - (Удаленая работа; Полная занятость) 

(Все в пределах нашей реальности, аномальных данных не выявленно)

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [79]:
query_5_1 = ''' select e.name as employer_name
                from employers as e 
                join vacancies as v on v.employer_id = e.id
                group by 1
                order by count(*) desc 
'''

In [80]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
display(df.iloc[0], df.iloc[4])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1179574961.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


employer_name    Яндекс
Name: 0, dtype: object

employer_name    Газпром нефть
Name: 4, dtype: object

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [20]:
query_5_2 = ''' select 
                    a.name as area_name,
                    count(e.id) as cnt_employers,
                    count(v.id) as cnt_vacancies
                from areas as a 
                left join vacancies as v on v.area_id = a.id
                left join employers as e on e.area = a.id
                group by 1
                having count(v.id) = 0
                order by 2 desc
'''


In [21]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
display(df.head())

C:\Users\97030\AppData\Local\Temp\ipykernel_18628\3510841220.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area_name  cnt_employers  cnt_vacancies
0              Россия            410              0
1           Казахстан            207              0
2  Московская область             75              0
3  Краснодарский край             19              0
4  Ростовская область             18              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [83]:
query_5_3 = ''' select 
                    e.name as emp_name,
                    count(distinct v.area_id) as cnt_areas
                from employers as e
                left join vacancies as v on v.employer_id = e.id
                group by 1
                order by 2 desc
'''


In [84]:
df = pd.read_sql_query(query_5_3, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1682211407.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


emp_name  cnt_areas
0                      Яндекс        181
1                  Ростелеком        152
2                  Спецремонт        116
3      Поляков Денис Иванович         88
4                   ООО ЕФИН          71
...                       ...        ...
23170                  СДЕЛКА          0
23171            Alandr Group          0
23172                     СДК          0
23173      Lemon Land Lombard          0
23174                  Махина          0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [85]:
query_5_4 = ''' select 
                    count(e.id) as cnt_employers 
                from employers as e 
                left join employers_industries as ei on ei.employer_id = e.id
                where ei.industry_id is null
''' 

In [86]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\375529716.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


cnt_employers
0           8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [87]:
query_5_5 = ''' select 
                    e.name as company 
                from employers as e 
                left join employers_industries as ei on ei.employer_id = e.id
                group by 1
                having count(ei.industry_id) = 4
                order by 1
'''

In [88]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
display(df.iloc[2])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1445855388.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


company    2ГИС
Name: 2, dtype: object

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [89]:
query_5_6 = ''' select count(*)
                from employers as e
                join employers_industries as ei on ei.employer_id = e.id
                join industries as i on i.id = ei.industry_id
                where i.name = 'Разработка программного обеспечения' 
'''

In [90]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1761488907.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [91]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')

tbody = soup.find('tbody')
cities =  tuple(map(lambda x: x.text, tbody.find_all('a')[5:]))
print(cities)


('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [92]:
query_5_7 = f''' 
                with part1 as 
                (select
                    a.name as area_name,
                    count(v.id) as cnt_vacancies
                from employers as e 
                join vacancies as v on v.employer_id = e.id
                join areas as a on a.id = v.area_id
                where (e.name = 'Яндекс') and (a.name in {cities})
                group by e.name, area_name
                order by 2)
                
                select * from part1 
                
                union all
                
                select 'Total', sum(part1.cnt_vacancies)
                from part1 
'''

In [93]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\4127784153.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


area_name  cnt_vacancies
0              Омск           21.0
1         Челябинск           22.0
2        Красноярск           23.0
3         Волгоград           24.0
4             Пермь           25.0
5            Казань           25.0
6    Ростов-на-Дону           25.0
7               Уфа           26.0
8            Самара           26.0
9         Краснодар           30.0
10          Воронеж           32.0
11      Новосибирск           35.0
12  Нижний Новгород           36.0
13     Екатеринбург           39.0
14  Санкт-Петербург           42.0
15           Москва           54.0
16            Total          485.0

***

### Выводы по анализу работодателей

- Яндекс и Газпром нефть - две очень крупной компании, соответсвенно можно предположить, что количесество персонала, которого необходимо для обеспечения деятельности компании, достаточно велико, следовательно, кол-во вакансий должно быть соответсвенно много, глядя на полученную выбоку данных, делаем вывод что мое предположение верно.

- Замечен странный феномен, в таблице areas (регионы) присутствуют название стран, что достаточно странно, ведь идет речь о регионах, а не о странах. Еще более странным является то, что в, так скажем, регионе "России", кол-во вакансий составляет нуль, но при этом кол-во работадателей 410

- Количество работадателей, у которых не указана сфера деятельнности равно 8419 (достаточно много)

- Количество работодателей, у которых в качестве сферы деятельности указана «Разработка программного обеспечения» равно 3553

- В каждом городе-миллионнике компания "Яндекс" предоставляет вакансии


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [94]:
query_6_1 = ''' select 
                    count(*) as cnt_data_vac
                from vacancies as v
                where  lower(v.name) like '%data%' or lower(v.name) like '%данн%'
'''

In [95]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\2931938894.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


cnt_data_vac
0          1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [96]:
query_6_2 = ''' select *
                from vacancies
                where  ((lower(name) like '%data scientist%') or
                        (lower(name) like '%data science%') or 
                        (lower(name) like '%исследователь данных%') or
                        ((upper(name) like '%ML%') and (upper(name) not like '%HTML%')) or 
                        (lower(name) like '%machine learning%') or
                        (lower(name) like '%машинн%обучен%')) AND
                        ((lower(name) like '%junior%') or
                        (experience = 'Нет опыта') or
                        (employment = 'Стажировка'))
'''

In [97]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
display(df.shape[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1596912114.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [98]:
query_6_3 = ''' select count(*)
                from vacancies
                where  ((lower(name) like '%data scientist%') or
                        (lower(name) like '%data science%') or 
                        (lower(name) like '%исследователь данных%') or
                        ((upper(name) like '%ML%') and (upper(name) not like '%HTML%')) or 
                        (lower(name) like '%machine learning%') or
                        (lower(name) like '%машинн%обучен%')) AND 
                        ((upper(key_skills) like '%SQL%') or 
                        (lower(key_skills) like '%postgres%'))
                        
'''

In [99]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
display(df.iloc[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\3265759844.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count    229
Name: 0, dtype: int64

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [100]:
query_6_4 = ''' select count(*)
                from vacancies
                where  ((lower(name) like '%data scientist%') or
                        (lower(name) like '%data science%') or 
                        (lower(name) like '%исследователь данных%') or
                        ((upper(name) like '%ML%') and (upper(name) not like '%HTML%')) or 
                        (lower(name) like '%machine learning%') or
                        (lower(name) like '%машинн%обучен%')) AND 
                        (upper(key_skills) like '%PYTHON%')
                        
'''

In [101]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
display(df.iloc[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1400223441.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count    357
Name: 0, dtype: int64

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [102]:
query_6_5 = ''' 
                with selection as(
                select 
                        (length(key_skills) -
                        length(replace(key_skills,CHR(9), ''))) + 1 as count_key_skills
                from vacancies
                where  ((lower(name) like '%data scientist%') or
                        (lower(name) like '%data science%') or 
                        (lower(name) like '%исследователь данных%') or
                        ((name like '%ML%') and (upper(name) not like '%HTML%')) or 
                        (lower(name) like '%machine learning%') or
                        (lower(name) like '%машинн%обучен%'))  AND
                        (key_skills is not Null) 
                )
                
                select round(avg(count_key_skills), 2)
                from selection               
'''

In [103]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
display(df.iloc[0])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\1530712187.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round    6.41
Name: 0, dtype: float64

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [104]:
display(df_vacancies.columns)

Index(['id', 'name', 'key_skills', 'schedule', 'experience', 'employment',
       'salary_from', 'salary_to', 'area_id', 'employer_id'],
      dtype='object')

In [105]:
query_6_6 = ''' 
                /* Основная выборка значений по которой будет происходить
                группировка данных*/
                with selection as(
                        
                /* Вакансии для DS без дополнительных условий,
                так скажем, шаблонная выборка */
                with DS_vac as (select *
                from vacancies
                where  ((lower(name) like '%data scientist%') or
                        (lower(name) like '%data science%') or 
                        (lower(name) like '%исследователь данных%') or
                        ((name like '%ML%') and (upper(name) not like '%HTML%')) or 
                        (lower(name) like '%machine learning%') or
                        (lower(name) like '%машинн%обучен%')) and (key_skills IS NOT NULL)
                ) -- КОНЕЦ ШАБЛОННОЙ ВЫБОРКИ
                
                
                /* Рассматриваем вакансии, где указана нижняя граница
                ИЛИ верхняя граница зарплатной вилки */
                (select 
                        experience ,
                        sum(coalesce(salary_from, 0) + coalesce(salary_to, 0)) as amount,
                        count(*) as cnt
                from DS_vac
                where   ((salary_from is not null) and (salary_to is null)) or
                        ((salary_from is null) and (salary_to is not null)) 
                group by 1)
                
                Union all
                
                /* Рассматриваем вакансии, где указана нижняя граница
                И верхняя граница зарплатной вилки */
                (select 
                        experience,
                        sum((salary_from + salary_to)/2),
                        count(*)
                from DS_vac
                where (salary_from is not null) and (salary_to is not null)
                group by 1)
                
                ) -- КОНЕЦ ОСНОВНОЙ ВЫБОРКИ
                
                /* Итоговый запрос */
                select
                        experience,
                        round(sum(amount) / sum(cnt)) as mean_salary
                from selection 
                group by 1 
'''

In [106]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
display(df[df.experience == 'От 3 до 6 лет'])

C:\Users\97030\AppData\Local\Temp\ipykernel_15772\3519224937.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  mean_salary
0  От 3 до 6 лет     243115.0

***

### Выводы по предметному анализу

- Всего вакансий, относящихся к данным - 1771
- Количество вакансий, подходящих для начинающего (junior) дата-сайнтиста - 51
- 229 вакансий для DS, в которых в качетве ключевого навыка цказан SQL или postgres
- Количество вакансий, в которых в качестве ключевого навыка указан Python - 357
- Всего ключевых навыков в среднем, которых указывают в вакансиях для DS - 6,41
- Зарплата, на которую в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет - 243115(руб)


# Общий вывод по проекту

### Дополнительные исследования данных 

In [67]:
query_7_1 = f"""select 
                    v.name as name,
                    round(avg(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0) as avg_salary
                from vacancies as v
                group by 1 
                having avg(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)) is not null
                order by 2 desc
                limit 10
"""

In [68]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_18628\2101136656.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


name  avg_salary
0                         СТО/Директор по разработке    750000.0
1                     Брокер по продаже апартаментов    750000.0
2              Frontend Developer in Unicorn Company    717000.0
3             FullStack Developer in Unicorn Company    717000.0
4                1С:Эксперт по производительности 1С    650000.0
5                    Руководитель команды разработки    600000.0
6                Senior DevOps in Blockchain Startup    587500.0
7  Senior Software Developer (upwork / иностранны...    555261.0
8                       Директор IT (HoReCa, Retail)    550000.0
9                           Project/Delivery manager    546087.0

In [69]:
query_7_2 = f"""select e.name as name,
                        round(avg(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0) as avg_salary
                from employers as e
                join vacancies as v on v.employer_id = e.id
                group by 1 
                having avg(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)) is not null
                order by 2 desc
                limit 10
"""

In [70]:
df = pd.read_sql_query(query_7_2, connection)
display(df)

C:\Users\97030\AppData\Local\Temp\ipykernel_18628\3686448377.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


name  avg_salary
0              Alpha Personnel   1000000.0
1                          ЦТП    650000.0
2                       RuWork    555261.0
3                Петухова Алия    552700.0
4                  BeRecruited    550000.0
5                       Пикабу    550000.0
6  Агентство недвижимости ЛЕТО    550000.0
7                Август, Фирма    500000.0
8                  ВСП Партнер    500000.0
9             StaffRecruitment    483333.0

### Подведем итоги дополнительного исследования 

- Рассмотрели 10 самых высокооплачиваемых вакансий в выбранном датасете
- Так же иссследовали 10 работадателей, у которых самые высокие зарплаты (учитывая все их вакансии)

# Итог исследования, обобщенные выводы

Подводя итоги исследовния, можно выделить основные ключевые этапы проделанной работы:
- Ознакомление с данными (Data Understanding)
- Предварительный анализ данных (EDA)
- Детальный анализ данных
- Анализ работадателей 
- Предметных анализ данных

Все эти этапы неотъемлемы для выполнения поставленной задачи (создание модели машинного обучения, которая будет рекомендовать вакансии клиентам агентства). 

В данном кейсе мы сделали исследования, в которых рассматривали вакансии Data Science. Понятное дело, что можно рассмотреть другие вакансии и найти в них другие закономерности, может быть даже аномальные данные.

Множество данных, рассматривыаемых в нашем проекте, не являются чем-то удивительным или универсальным, они являются приемлемыми для нашей реальности. Однако были замечены возможные выбросы, такие как в некоторых регоинах нет ни одной вакансии. Например: В России ноль вакансий, разве это возможно? Стоит перепроверить данные и провести дополнительное исследование.

В исследовании были исследованы различные данные для того, чтобы найти закономерности, что очень важно для дальшейшей поставленной задачи. Как всем известно, одним из самых важным ориентиром отклика на вакансию является заработная плата/заработная вилка, поэтому мы не могли, не рассмотреть этот важный ориентир в нашем исследовании.